### ISSUE TITLE
Eq1, Ord1, Show1, Read1 with ghc8

### ISSUE THREAD

**Issue by reuleaux** [ekmett/bound#37]
The following example works fine for me with ghc7.10.3, but not with ghc8. It is complaining

```
src/Exp.hs:56:22: error:
    • No instance for (Data.Functor.Classes.Show1 (Exp t))
        arising from the second field of ‘Lam’ (type ‘Scope () (Exp t) a’)
      Possible fix:
        use a standalone 'deriving instance' declaration,
          so you can specify the instance context yourself
    • When deriving the instance for (Show (Exp t a))
```

and similar for `Eq1`, `Ord1`, `Read1`

Sorry, if I am asking the obvious, but I would appreciate some help of how to
derive `Eq1`, `Ord1`, `Show1`, `Read1` for `(Exp t)`

Note that the use of two type variables: `t`, `a` makes my expressions: `Exp t a` a little more
complicated than most of the Bound examples I usually see (with just `Exp a`), an idea I got by 
cheating at `Term`s in the ermine code originally:

`https://github.com/ermine-language/ermine/blob/master/src/Ermine/Syntax/Term.hs`

 but they have worked fine for me so far (and I have come to rely on them).

Not sure if I have a chance to apply a simple/automatic solution like

```haskell
import Data.Eq.Deriving (deriveEq1)
import Text.Show.Deriving (deriveShow1)
```
from `deriving-compat` relying on Template Haskell, as in the example at
`https://github.com/ekmett/bound`

at all, given my two expressions of `t` and `a`, but I would be happy already
if I just got my code compiled with ghc8

I have tried `deriving instance Show t => Show1 (Exp t)` as well, with no luck so far.

Im am using the latest version of `Bound` from github (not the slighly older version on hackage),
cf. `https://github.com/ekmett/bound/issues/32`, ie. `Bound` itself compiles fine for me
under ghc8 from debian sid/unstable.

Thanks in advance,
  Andreas

```haskell

{-# LANGUAGE DeriveFunctor #-}
{-# LANGUAGE DeriveFoldable #-}
{-# LANGUAGE DeriveTraversable #-}


{-# LANGUAGE StandaloneDeriving #-}

{-# LANGUAGE CPP #-}


module Exp where


import Control.Monad



import Prelude.Extras

import Bound


infixl 9 :@


data Exp t a =

  V a

  | (Exp t a) :@ (Exp t a)

  | Lam  t       (Scope () (Exp t) a)



#if MIN_VERSION_GLASGOW_HASKELL(8,0,1,0)

-- what to do here for ghc8 ? 


-- these lines just copied, but don't work
  deriving (Eq, Ord, Show, Read, Functor, Foldable, Traversable)

instance Eq t => Eq1 (Exp t)
instance (Ord t) => Ord1 (Exp t)
instance Show t => Show1 (Exp t)
instance Read t => Read1 (Exp t)



#else
-- works fine in ghc 7.10.3

  deriving (Eq, Ord, Show, Read, Functor, Foldable, Traversable)

instance Eq t => Eq1 (Exp t)
instance (Ord t) => Ord1 (Exp t)
instance Show t => Show1 (Exp t)
instance Read t => Read1 (Exp t)

#endif



instance Applicative (Exp t) where
  pure = V
  (<*>) = ap



instance Monad (Exp t) where

  return = V

  
  V a        >>= f = f a

  (x :@ y)   >>= f = (x >>= f) :@ (y >>= f)

  Lam n s    >>= f = Lam n (s >>>= f)



-- | 
-- >>> lam "a" $ V "a" :@ V "b"
-- Lam "a" (Scope (V (B ()) :@ V (F (V "b"))))



lam v b = Lam v (abstract1 v b)


```

**Comment by phadej** [comment]
Haven't you checked the example in readme.md?

It has

```hs
import Data.Eq.Deriving (deriveEq1)      -- these two are from the
import Text.Show.Deriving (deriveShow1)  -- deriving-compat package

deriveEq1 ''Exp
deriveShow1 ''Exp
```

We have to go thru all example to use this approach.

**Comment by reuleaux** [comment]
Well yes, I mentioned that already: things are more complicated in my
case of Exp depending on two type vars `t` and `a`. Basically I would 
need something like

```hs
deriveEq1 ''(Exp t)
deriveShow1 ''(Exp t)
```

my `(Exp t)` corresponding to the simple expression cited in the example (expecting
one more `a` param). 


```hs
''(Exp t)
```

doesn't work however. Maybe there is some way to make this work, that I
am just not aware of, some TH magic perhaps? I would be happy to
learn then. 

Thanks for taking the time and coming up with some idea anyway.
-Andreas

**Comment by jakzale** [comment]
I can compile the example form README.md on ghc 8.0.1 by using deriving-compat to derive both the `Eq1` and `Eq` instances, and `Show1` and `Show` instances.

```haskell
deriveEq1 ''Exp
deriveEq ''Exp
deriveShow1 ''Exp
deriveShow ''Exp
```

**Comment by reuleaux** [comment]
Thanks a lot, the example in the README.md is another, simpler one,
though. As explained already: my expressions take two type parameters: t
and a.

**Comment by RyanGlScott** [comment]
@reuleaux: You were 90% of the way to getting it to compile in https://github.com/ekmett/bound/issues/37#issuecomment-258147317. You were just using Template Haskell incorrectly.

All you need to do is this:

```haskell
import Data.Eq.Deriving (deriveEq1)
import Text.Show.Deriving (deriveShow1)

data Exp t a =
  V a
  | (Exp t a) :@ (Exp t a)
  | Lam  t       (Scope () (Exp t) a)

deriveEq1 ''Exp
deriveShow1 ''Exp
```

That's it. `deriveEq1` and friends take a Template Haskell type constructor `Name` as an argument, which is syntactically denoted with two single quotes followed by a tycon. Something like `deriveEq ''(Exp t)` doesn't make sense, since `(Exp t)` isn't a plain tycon, and in any case, `deriving-compat` is smart enough to work with datatypes with multiple type parameters!

**Comment by reuleaux** [comment]
@RyanGlScott thanks a lot. yes, this seems to do the trick. I had a a brief look only so far:
seems that I need standalone deriving instances then like so, but fine:

```hs
deriveEq1 ''Exp
deriveOrd1 ''Exp
deriveShow1 ''Exp
deriveRead1 ''Exp



deriving instance (Eq t, Eq a)     => Eq (Exp t a)
deriving instance (Ord t, Ord a)   => Ord (Exp t a)
deriving instance (Show t, Show a) => Show (Exp t a)
deriving instance (Read t, Read a) => Read (Exp t a)
```

I will want to work through a few more examples still, and fingers crossed: hope that the above will work in all cases, will post a complete working solution then.

**Comment by phadej** [comment]
@reuleaux updates on this issue, is it resolved for you?

**Comment by reuleaux** [comment]
Hi Oleg,

yes, it is resolved for me.

Thanks for asking, and sorry for not getting back to you earlier:
I meant to write up some bits about my experience in this regard,
but didn't get a chance to, and am now, somehow unforseen, busy in
my mother's apartment in Munich, thus distracted again.

The interested reader / programmer facing similar issues
may thus be referred to my Pire code:

  https://github.com/reuleaux/pire

and in particular the code in Syntax/Expr.hs:

  https://github.com/reuleaux/pire/blob/master/src/Pire/Syntax/Ehttps://github.com/reuleaux/pire/blob/master/src/Pire/Syntax/Expr.hsxpr.hs

Things that I remember:

* Given that the data structures for
  expressions/annotations/arguments etc in that module are mutually recursive,
  I had group all the deriveEq1/deriveOrd1/deriveShow1 code at the
  the very end of the module

* deriving worked fine for me for simple things like pairs / lists /
  lists of pairs etc, but did not for triples (at least not at the time
  when I did the coding, maybe the situation has improved now?).
  Thus I had to handle deriving for triples: (,,), or  lists of triples
  for that matter, myself:

    Eq3, Show3, liftShowsPrec3, liftShowList3 etc.

  thereby cheating at the code for pairs:

    https://hackage.haskell.org/package/base-4.9.0.0/docs/src/Data.Functor.Classes.html

  In any case: I would think, that triples are not that exoctic in
  Haskell code, and the more cases like these are covered by the
  transformers lib, the more likely will using the deriving mechanism be
  a smooth "works out of the box" experience.

* What else? I am sure I have forgotten something.

Anyway, I would think that the issue can be closed, and thanks again.

Andreas
